File: back/src/whombat/routes/sound_events.py
"""REST API routes for sound events."""

from uuid import UUID

from fastapi import APIRouter, Depends

from whombat import api, schemas
from whombat.filters.sound_events import SoundEventFilter
from whombat.routes.dependencies import Session
from whombat.routes.types import Limit, Offset

__all__ = [
    "sound_events_router",
]

sound_events_router = APIRouter()


@sound_events_router.get(
    "/detail/recording/",
    response_model=schemas.Recording,
)
async def get_sound_event_recording(
    session: Session,
    sound_event_uuid: UUID,
):
    """Get the recording for a sound_event."""
    sound_event = await api.sound_events.get(session, sound_event_uuid)
    return await api.sound_events.get_recording(session, sound_event)


@sound_events_router.get(
    "/",
    response_model=schemas.Page[schemas.SoundEvent],
)
async def get_sound_events(
    session: Session,
    limit: Limit = 10,
    offset: Offset = 0,
    sort_by: str = "-created_on",
    filter: SoundEventFilter = Depends(SoundEventFilter),  # type: ignore
):
    """Get a page of sound events."""
    datasets, total = await api.sound_events.get_many(
        session,
        limit=limit,
        offset=offset,
        filters=[filter],
        sort_by=sort_by,
    )
    return schemas.Page(
        items=datasets,
        total=total,
        offset=offset,
        limit=limit,
    )


@sound_events_router.post(
    "/",
    response_model=schemas.SoundEvent,
)
async def create_sound_event(
    session: Session,
    recording_uuid: UUID,
    data: schemas.SoundEventCreate,
):
    """Create a sound_event."""
    recording = await api.recordings.get(session, recording_uuid)
    sound_event = await api.sound_events.create(
        session,
        recording,
        data.geometry,
    )
    await session.commit()
    return sound_event


@sound_events_router.get(
    "/detail/",
    response_model=schemas.SoundEvent,
)
async def get_sound_event(
    session: Session,
    sound_event_uuid: UUID,
):
    """Get a sound_event."""
    return await api.sound_events.get(session, sound_event_uuid)


@sound_events_router.patch(
    "/detail/",
    response_model=schemas.SoundEvent,
)
async def update_sound_event(
    session: Session,
    sound_event_uuid: UUID,
    data: schemas.SoundEventUpdate,
):
    """Update a sound_event."""
    sound_event = await api.sound_events.get(session, sound_event_uuid)
    response = await api.sound_events.update(session, sound_event, data)
    await session.commit()
    return response


@sound_events_router.post(
    "/detail/features/",
    response_model=schemas.SoundEvent,
)
async def add_sound_event_feature(
    session: Session,
    sound_event_uuid: UUID,
    name: str,
    value: float,
):
    """Add a feature to a sound_event."""
    sound_event = await api.sound_events.get(session, sound_event_uuid)
    feature = await api.features.get_feature(session, name=name, value=value)
    response = await api.sound_events.add_feature(
        session,
        sound_event,
        feature,
    )
    await session.commit()
    return response


@sound_events_router.delete(
    "/detail/features/",
    response_model=schemas.SoundEvent,
)
async def remove_sound_event_feature(
    session: Session,
    sound_event_uuid: UUID,
    name: str,
    value: float,
):
    """Remove a feature from a sound_event."""
    sound_event = await api.sound_events.get(session, sound_event_uuid)
    feature = await api.features.get_feature(session, name=name, value=value)
    response = await api.sound_events.remove_feature(
        session,
        sound_event,
        feature,
    )
    await session.commit()
    return response


@sound_events_router.patch(
    "/detail/features/",
    response_model=schemas.SoundEvent,
)
async def update_sound_event_feature(
    session: Session,
    sound_event_uuid: UUID,
    name: str,
    value: float,
):
    """Update a feature on a sound_event."""
    sound_event = await api.sound_events.get(session, sound_event_uuid)
    feature = await api.features.get_feature(session, name=name, value=value)
    response = await api.sound_events.update_feature(
        session,
        sound_event,
        feature,
    )
    await session.commit()
    return response

File: front/src/lib/api/sound_events.ts
import { AxiosInstance } from "axios";

import { GetMany, Page } from "@/lib/api/common";
import * as schemas from "@/lib/schemas";
import type * as types from "@/lib/types";

const DEFAULT_ENDPOINTS = {
  getMany: "/api/v1/sound_events/",
  get: "/api/v1/sound_events/detail/",
  getRecording: "/api/v1/sound_events/detail/recording/",
  update: "/api/v1/sound_events/detail/",
  delete: "/api/v1/sound_events/detail/",
  create: "/api/v1/sound_events/",
  addFeature: "/api/v1/sound_events/detail/features/",
  updateFeature: "/api/v1/sound_events/detail/features/",
  removeFeature: "/api/v1/sound_events/detail/features/",
};

export function registerSoundEventAPI(
  axios: AxiosInstance,
  endpoints: typeof DEFAULT_ENDPOINTS = DEFAULT_ENDPOINTS,
) {
  async function get(uuid: string): Promise<types.SoundEvent> {
    const { data } = await axios.get(endpoints.get, {
      params: { sound_event_uuid: uuid },
    });
    return schemas.SoundEventSchema.parse(data);
  }

  async function getMany(
    query: types.GetMany & types.SoundEventFilter,
  ): Promise<types.Page<types.SoundEvent>> {
    const params = GetMany(schemas.SoundEventFilterSchema).parse(query);
    const { data } = await axios.get(endpoints.getMany, {
      params: {
        limit: params.limit,
        offset: params.offset,
        sort_by: params.sort_by,
        geometry_type__eq: params.geometry_type,
        created_on__before: params.created_on?.before,
        created_on__after: params.created_on?.after,
        created_on__on: params.created_on?.on,
        recording__eq: params.recording?.uuid,
        feature__name: params.feature?.name,
        feature__gt: params.feature?.gt,
        feature__lt: params.feature?.lt,
      },
    });
    return Page(schemas.SoundEventSchema).parse(data);
  }

  async function getRecording(
    soundEvent: types.SoundEvent,
  ): Promise<types.Recording> {
    const { data } = await axios.get(endpoints.getRecording, {
      params: { sound_event_uuid: soundEvent.uuid },
    });
    return schemas.RecordingSchema.parse(data);
  }

  async function createSoundEvent(
    recording: types.Recording,
    data: types.SoundEventCreate,
  ): Promise<types.SoundEvent> {
    const body = schemas.SoundEventCreateSchema.parse(data);
    const { data: responseData } = await axios.post(endpoints.create, body, {
      params: { recording_uuid: recording.uuid },
    });
    return schemas.SoundEventSchema.parse(responseData);
  }

  async function updateSoundEvent(
    soundEvent: types.SoundEvent,
    data: types.SoundEventUpdate,
  ): Promise<types.SoundEvent> {
    const body = schemas.SoundEventUpdateSchema.parse(data);
    const { data: responseData } = await axios.patch(endpoints.update, body, {
      params: { sound_event_uuid: soundEvent.uuid },
    });
    return schemas.SoundEventSchema.parse(responseData);
  }

  async function deleteSoundEvent(
    soundEvent: types.SoundEvent,
  ): Promise<types.SoundEvent> {
    const { data } = await axios.delete(endpoints.delete, {
      params: {
        sound_event_uuid: soundEvent.uuid,
      },
    });
    return schemas.SoundEventSchema.parse(data);
  }

  async function addFeature(
    soundEvent: types.SoundEvent,
    feature: types.Feature,
  ) {
    const { data } = await axios.post(
      endpoints.addFeature,
      {},
      {
        params: {
          sound_event_uuid: soundEvent.uuid,
          name: feature.name,
          value: feature.value,
        },
      },
    );
    return schemas.SoundEventSchema.parse(data);
  }

  async function updateFeature(
    soundEvent: types.SoundEvent,
    feature: types.Feature,
  ) {
    const { data } = await axios.patch(
      endpoints.updateFeature,
      {},
      {
        params: {
          sound_event_uuid: soundEvent.uuid,
          name: feature.name,
          value: feature.value,
        },
      },
    );
    return schemas.SoundEventSchema.parse(data);
  }

  async function removeFeature(
    soundEvent: types.SoundEvent,
    feature: types.Feature,
  ) {
    const { data } = await axios.delete(endpoints.removeFeature, {
      params: {
        sound_event_uuid: soundEvent.uuid,
        name: feature.name,
        value: feature.value,
      },
    });
    return schemas.SoundEventSchema.parse(data);
  }

  return {
    get,
    getMany,
    create: createSoundEvent,
    update: updateSoundEvent,
    delete: deleteSoundEvent,
    addFeature,
    getRecording,
    updateFeature,
    removeFeature,
  } as const;
}
